services/queue/src/api.js

```diff
@@ -300,6 +300,132 @@ builder.declare({
   return res.reply(result);
 });
 
+/** Cancel all tasks in a group */
+builder.declare({
+  method: 'post',
+  route: '/task-group/:taskGroupId/cancel',
+  name: 'cancelTaskGroup',
+  scopes: {
+    AnyOf: [
+      'queue:cancel-task-group:<schedulerId>/<taskGroupId>',
+      'queue:cancel-task-in-project:<projectId>',
+    ],
+  },
+  stability: APIBuilder.stability.experimental,
+  category: 'Tasks',
+  input: undefined,
+  output: 'cancel-task-group-response.yml',
+  title: 'Cancel Task Group',
+  description: [
+    'This method will cancel all tasks with a given `taskGroupId` that are not resolved yet.',
+    'That means all tasks in either `unscheduled`, `pending` or `running` states.',
+    'Behaviour is similar to the `cancelTask` method.',
+    '',
+    '**Remark** this operation does not guarantee that new tasks created for this `taskGroupId`',
+    'would automatically be rejected. For example when running task keeps creating new tasks.',
+  ].join('\n'),
+}, async function(req, res) {
+  const taskGroupId = req.params.taskGroupId;
+  // allow query.reasonResolved ? cancel | superseded
+
+  const [taskGroup] = await this.db.fns.get_task_group(taskGroupId);
+  if (!taskGroup) {
+    return res.reportError('ResourceNotFound',
+      'No task-group with taskGroupId: `{{taskGroupId}}`', {
+        taskGroupId,
+      },
+    );
+  }
+
+  // in order to authorize call we would need to fetch a task to retrieve schedulerId and projectId
+  const someTaskInGroup = Task.fromDbRows(
+    await this.db.fns.get_tasks_by_task_group_projid(taskGroupId, 1, 0),
+  );
+  if (!someTaskInGroup) {
+    return res.reportError('ResourceNotFound',
+      'No tasks found for task-group with taskGroupId: `{{taskGroupId}}`', {
+        taskGroupId,
+      },
+    );
+  }
+
+  await req.authorize({
+    taskGroupId,
+    schedulerId: someTaskInGroup.schedulerId,
+    projectId: someTaskInGroup.projectId,
```

Comment:
Is it a requirement that all tasks in the same taskgroup have the same projectId? I'm pretty confident the same schedulerId is required, but I didn't know `projectId` had to be the same.

Reply:
I don't think there is a strict rule, code doesn't enforce all tasks to have same projectId in the group. But according to docs this is the mechanism to set up the permission system. In the call above `taskGroupId`, `schedulerId` and `projectId` would be injected in scope templates:
```
 'queue:cancel-task-group:<schedulerId>/<taskGroupId>',
 'queue:cancel-task-in-project:<projectId>',
```
to see if current user has permissions. 

This mimics the behavior of the `cancelTask` method which uses similar scopes:
```
    'queue:cancel-task:<schedulerId>/<taskGroupId>/<taskId>',
    'queue:cancel-task-in-project:<projectId>',
```

Reply:
So the issue here is that the code takes an arbitrary task from the task group to see what projectId it has. If scopes are sufficient for _that_ task, then the transaction is approved for _all_ tasks in that task group, even if the client that calls this endpoint doesn't have the scopes to delete all the tasks. This should be changed to return the unique projectIds, and require `queue:cancel-task-in-project:<project>` for _all_ projectIds that exist in the set of tasks that it would like to cancel. This is one of the reasons I am wary of implementing this directly in the database, and not gating it on the `cancelTask` implementation (which would ensure consistent scope treatment).